### ISSUE TITLE
Missing ArtifactStateHelper extension causes NullPointerException errors and unnecessary syncs

### ISSUE THREAD

**Issue by alexhanson** [bazelbuild/intellij#2874]
**Short version:** None of the checked-in XML configurations for the plugin define an `<ArtifactStateHelper>` extension, which seems to prevent the plugin from using cached metadata when opening a project. Modifying the plugin to load `LocalArtifactStateProtoConverter` seems to fix the problem.

### Backstory
After upgrading from version `2020.01.28.0.3` of the plugin to recent versions, I noticed that opening a Bazel project in IntelliJ results in a lot more Bazel activity than before. The same problem occurs for plugin versions `2021.07.27.1.0-api-version-211` and `2021.08.10.0.2-api-version-211`, as published to the JetBrains Marketplace.

I upgraded the plugin alongside an upgrade of IntelliJ itself. I'm now using IntelliJ IDEA Ultimate 2021.1.3 with a Java-centric Bazel project.

I've done multiple rounds of opening the project in IntelliJ, waiting for all indexing/syncing to settle, closing the project, and opening it again. I also tried clearing IntelliJ caches and running `bazel clean` commands. The indexing/syncing process takes just as long each time, with Bazel spending most of the time on an incremental sync (and maybe two).

When I watch the IntelliJ application logs during that window, I see stack traces including `NullPointerException` errors like this:

<details>
<summary>Stack trace during project open</summary>

```
java.lang.NullPointerException: Null file
	at com.google.idea.blaze.java.sync.jdeps.AutoValue_JdepsState_JdepsData.<init>(AutoValue_JdepsState_JdepsData.java:32)
	at com.google.idea.blaze.java.sync.jdeps.JdepsState$JdepsData.create(JdepsState.java:50)
	at com.google.idea.blaze.java.sync.jdeps.JdepsState.lambda$fromNewProto$3(JdepsState.java:74)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at com.google.idea.blaze.java.sync.jdeps.JdepsState.fromNewProto(JdepsState.java:78)
	at com.google.idea.blaze.java.sync.jdeps.JdepsState.fromProto(JdepsState.java:84)
	at com.google.idea.blaze.java.sync.jdeps.JdepsState.access$100(JdepsState.java:39)
	at com.google.idea.blaze.java.sync.jdeps.JdepsState$Extractor.extract(JdepsState.java:171)
	at com.google.idea.blaze.java.sync.jdeps.JdepsState$Extractor.extract(JdepsState.java:167)
	at com.google.idea.blaze.base.model.SyncData.lambda$extract$0(SyncData.java:39)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at com.google.idea.blaze.base.model.SyncData.extract(SyncData.java:41)
	at com.google.idea.blaze.base.model.SyncState.fromProto(SyncState.java:61)
	at com.google.idea.blaze.base.model.BlazeProjectData.fromProto(BlazeProjectData.java:82)
	at com.google.idea.blaze.base.model.BlazeProjectData.loadFromDisk(BlazeProjectData.java:151)
	at com.google.idea.blaze.base.sync.data.BlazeProjectDataManagerImpl.loadProject(BlazeProjectDataManagerImpl.java:85)
	at com.google.idea.blaze.base.sync.data.BlazeProjectDataManagerImpl.loadProjectRoot(BlazeProjectDataManagerImpl.java:71)
	at com.google.idea.blaze.base.prefetch.PrefetchProjectInitializer.getBlazeProjectData(PrefetchProjectInitializer.java:101)
	at com.google.idea.blaze.base.prefetch.PrefetchProjectInitializer.lambda$prefetchProjectFiles$1(PrefetchProjectInitializer.java:81)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleAsyncTask.runInterruptibly(TrustedListenableFutureTask.java:160)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleAsyncTask.runInterruptibly(TrustedListenableFutureTask.java:143)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:69)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:668)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:665)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:665)
	at java.base/java.lang.Thread.run(Thread.java:829)
```
</details>

### Root cause
Digging into the `NullPointerException` above, I discovered that the null value in question was supposed to be an `ArtifactState` object, derived from paths in the `cache.dat.gz` file for this IntelliJ project. Investigating further, I found that the plugin never enters [this loop](https://github.com/bazelbuild/intellij/blob/v2021.08.09/base/src/com/google/idea/blaze/base/filecache/ArtifactStateProtoConverter.java#L40) in `ArtifactStateProtoConverter` because no `<ArtifactStateHelper>` extensions are present.

It looks like the `<ArtifactStateHelper>` extension point was introduced in commit 1af8d0a996611d8e9b295c8b6cd7ca94dcffc630, first included in the `v2020.07.13` tagged release. At the time, the only `<ArtifactStateHelper>` directive referenced a class that seems internal to Google. That caused problems for external users, which was reported in issue #2014 and fixed in commit 9f65fdd190a79bebb0550d8988ddaad4bc12396f, first included in the `v2020.08.26` tagged release.

That first commit did include a `LocalArtifactStateProtoConverter` implementation of `ArtifactStateProtoConverter`, which is still present in the codebase. It doesn't appear that implementation was ever referenced in a checked-in `<ArtifactStateHelper>` directive, however.

### Prospective fix
I made a local build of the plugin after modifying `base/src/META-INF/blaze-base.xml` as follows:

```diff
diff --git a/base/src/META-INF/blaze-base.xml b/base/src/META-INF/blaze-base.xml
index 7d2f7722..666fc205 100644
--- a/base/src/META-INF/blaze-base.xml
+++ b/base/src/META-INF/blaze-base.xml
@@ -542,6 +542,7 @@
     <VcsSyncListener implementation="com.google.idea.blaze.base.sync.libraries.ExternalLibraryManager$VcsListener"/>
     <SettingsUiContributor implementation="com.google.idea.blaze.base.settings.ui.BlazeUserSettingsConfigurable$UiContributor" order="first" id="base"/>
     <MacroTargetProvider implementation="com.google.idea.blaze.base.query.BlazeQueryMacroTargetProvider" order="last"/>
+    <ArtifactStateHelper implementation="com.google.idea.blaze.base.filecache.LocalArtifactStateProtoConverter"/>
     <BinaryPathRemapper implementation="com.google.idea.blaze.base.async.process.MacBinaryPathRemapper" order="last"/>
     <CustomFormatter implementation="com.google.idea.blaze.base.buildmodifier.BuildifierCustomFormatter"/>
   </extensions>

```

Opening my project now involves very little work from Bazel itself, with this process completing in about a second once more:

```
Syncing project: Sync Project (startup)...
Updating in-memory state...
Sync finished
```

**Comment by alexhanson** [comment]
Upon further research, I see this bug was also reported in issue #2097 from September 2020.

**Comment by alexhanson** [comment]
I suspect issue #2392 from February 2021 is a reflection of this bug too.

**Comment by AlexeyGy** [comment]
Thank you very much for investigating and suggesting this. I think we dropped the ball here and it makes total sense to me that we fix this as this impacts quite a few projects (almost everyone using protobuf). I'll discuss this with the team and hopefully we can get this out asap.

I wonder if there is any way to require at least one implementation for a given extension point at compile time to prevent issues like this in the future. Of course can do it at runtime, via throwing an exception, that's surely better than an opaque NPE, but this wouldn't help plugin users very much. 

One idea to prevent this at compile time that I have would be to extend the [`intellij_plugin_library`](https://github.com/bazelbuild/intellij/blob/master/build_defs/intellij_plugin.bzl#L81) rule that we have .

**Comment by alexhanson** [comment]
Thanks for merging a fix, @AlexeyGy! Much appreciated 🙂

If you're looking to tidy up some other issues, I think these can all be closed now: #2095, #2097, #2220, #2392
